components/VideoPlayer.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+import React, { useEffect, useRef } from 'react';
+import { Alert } from 'react-native';
+import { Audio, Video } from 'expo-av';
+import { observer } from 'mobx-react';
+
+import { useStores } from '../hooks/useStores';
+import MediaTypes from '../constants/MediaTypes';
+import { msToTicks } from '../utils/Time';
+
+const VideoPlayer = observer(() => {
+	const { rootStore } = useStores();
+
+	const player = useRef(null);
+
+	// Set the audio mode when the video player is created
+	useEffect(() => {
+		Audio.setAudioModeAsync({
+			interruptionModeAndroid: Audio.INTERRUPTION_MODE_ANDROID_DO_NOT_MIX,
+			interruptionModeIOS: Audio.INTERRUPTION_MODE_IOS_DO_NOT_MIX,
+			playsInSilentModeIOS: true
+		});
+	}, []);
+
+	// Update the player when media type or uri changes
+	useEffect(() => {
+		if (rootStore.mediaStore.type === MediaTypes.Video) {
+			rootStore.didPlayerCloseManually = true;
+			player.current?.loadAsync({
+				uri: rootStore.mediaStore.uri
+			}, {
+				positionMillis: rootStore.mediaStore.positionMillis,
+				shouldPlay: true
+			});
+		}
+	}, [ rootStore.mediaStore.type, rootStore.mediaStore.uri ]);
+
+	// Update the play/pause state when the store indicates it should
+	useEffect(() => {
+		if (rootStore.mediaStore.shouldPlayPause) {
+			if (rootStore.mediaStore.isPlaying) {
+				player.current?.pauseAsync();
+			} else {
+				player.current?.playAsync();
+			}
+			rootStore.mediaStore.shouldPlayPause = false;
+		}
+	}, [ rootStore.mediaStore.shouldPlayPause ]);
+
+	// Close the player when the store indicates it should stop playback
+	useEffect(() => {
+		if (rootStore.mediaStore.shouldStop) {
+			rootStore.didPlayerCloseManually = false;
+			player.current?.dismissFullscreenPlayer()
+				.catch(console.debug);
+			rootStore.mediaStore.shouldStop = false;
+		}
+	}, [ rootStore.mediaStore.shouldStop ]);
+
+	return (
+		<Video
+			ref={player}
+			usePoster
+			posterSource={{ uri: rootStore.mediaStore.posterUri }}
+			resizeMode='contain'
+			useNativeControls
+			onReadyForDisplay={() => {
+				player.current?.presentFullscreenPlayer()
+					.catch(console.debug);
```

Comment:
Wouldn't it be good to have some message here for the user? Or is that taken care of by jf-web noticing that playback failed?

Reply:
Good call! Playback errors are handled in the `onError` function, but I added this to catch errors thrown when showing the fullscreen player when it is already being shown. I added some local state to debounce calls to present and dismiss and added an Alert dialog if an error happens when presenting.